core/queryalgebra/evaluation/src/main/java/org/eclipse/rdf4j/query/algebra/evaluation/impl/StandardQueryOptimizerPipeline.java

```diff
@@ -56,9 +56,9 @@ public Iterable<QueryOptimizer> getOptimizers() {
 				new UnionScopeChangeOptimizer(),
 				new QueryModelNormalizer(),
 				new ProjectionRemovalOptimizer(), // Make sure this is after the UnionScopeChangeOptimizer
+				new FilterOptimizer(),
 				new QueryJoinOptimizer(evaluationStatistics),
 				new IterativeEvaluationOptimizer(),
-				new FilterOptimizer(),
```

Comment:
This isn't _necessary_ for the fix, but pushing filter conditions down before doing join optimization reduces the number of tree walks the join optimizer potentially has to do.